Use native bits and registers for `CircuitData`.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

The following commits add native bits and register usage to `CircuitData`. This is the first public attempt at implementing these structures and is open to feedback.

### Details and comments

As we continue to move towards having more rust-native components as part of our core data model, we need to continue to promote more functionality to be able to work without needing intervention from Python. #13269 describes the cost paid whenever an instance of `CircuitData` and `DAGCircuit` needs to be constructed from Rust. There is a constant need for calling back to Python to create instances of `Qubit` and `Clbit` as needed.

An important point made towards this is to transfer the ownership of `Bits` and `Registers` to the circuit, but also to keep the current functionality of this API. Meaning that, instead of removing the dependency completely, we had to shift the source of truth to rust. The following commits achieve this by performing the following:

#### Adding representations of `QuantumRegister` and `ClassicalRegister` in Rust.

A simple representation using an `IndexSet`, and attributes such as `name` and `size`.

#### Modifying `BitData` currently `NewBitData` to also store register information and dynamically initialize the Python counter parts as requested:

- Registers are now kept in a `Vec<_>` and can be accessed by index or by key (`RegisterKey`).
- Bits are still referred to as either `Qubit(u32)` and `Clbit(u32)` and there exists a mapping between Python `Bits` and the specified index type.
    - The mapping has to be inside of an `RWLock` due to the dynamic initialization of Python `Bit`s.
- A new attribute: `BitInfo` keeps a list, organized by the bit's index, which keeps track of:
    - Whether a bit was added by a register into the circuit or not.
    - Which registers and which index within said register(s) does that bit belong in.
- Separate python specific and rust methods:
    - Python methods will perform dynamic initialization of bits and registers in Python when needed.
- Both `Bit`s and `Register`s from Python are stored as a `OnceLock` which gets dynamically initialized upon request based on the information stored here.
#### Changes to `CircuitData`

A new Python-free constructor `CircuitData::new()` was created to allow initialization of Rust instances without any need for interaction with Python unless otherwise requested. 
- Modifications were also added to the `CircuitData::with_capacity` constructor to also add the bits dynamically without accessing Python.
    - This constructor will still need a `py` token due to assignments to the `ParameterTable`.

### Known issues and drawbacks
- `copy()` can take up to twice as long to perform a copy of the circuit compared to previous iterations. This most likely due re-mapping of the registers.
- Risks with using `OnceLock` and `RWLock` as these can panic under the wrong circumstances.
- Python `Bit`s are fully opaque to the Rust ones and there's no validation other than the mapping and information stored in Rust.
- `DAGCircuit` does not use any of the new additions.
    - This is mostly due to the robust nature of the changes and a follow-up PR will add these changes to the `DAGCircuit`.
- Testing in Rust does not allow us to check if conversion works as needed into a `QuantumCircuit`.

### Tasks:
- [x] Implement a new `BitData` struct to store registers.
- [x] Implement the new `BitData` in `CircuitData`.
- [x] Fix all the broken tests.
- [x] Add rust-native tests to check if this is working properly.
- [ ] Implement partial python cross testing.

### Questions or comments?
Feel free to leave a review with any questions or concerns you might have.

raynelfss committed:
Initial: Register Infrastructure in Rust

raynelfss committed:
Add: `bits` method for register.
- Add missing traits on `BitInfo`

raynelfss committed:
Use interners for registry

raynelfss committed:
Fix: Add functional python `Bits`

raynelfss committed:
Initial Python registers

raynelfss committed:
Finalize Python Registers, prepare for replacements.

raynelfss committed:
Fix serialization in registers
- Fix bit display property.

raynelfss committed:
`BitData` refactor: Store Registers

raynelfss committed:
Add new `BitData` to `CircuitData`

raynelfss committed:
Add more rust native connections to `CircuitData` and `QuantumCircuit`.

raynelfss committed:
Fix: Do not re-add bits when adding registers.

raynelfss committed:
Fix: Incorrect addition of register cache

raynelfss committed:
Fix: Incorrect serialization and mutability
- Some functions in `CircuitData` had become mutable unnecessarily due to erroneous assumptions on the `BitData` `OnceCell` usage.
- Fix incorrect serialization to now include registers in `CircuitData::__reduce__`.
- Also add `CircuitData::__setstate__` to deal with the new register state.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit_u32_sequel

raynelfss committed:
Refactor `BitInfo` to account for bits added from registers.
- Fix bug with `BlueprintCircuit` that discarded most instances of `QuantumRegister`.
- Rename `RegisterAsKey::index()` to size, since it was erroneously named.

to24toro commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,65 @@
+use std::fmt::Debug;
+
+/// Keeps informatyion about where a bit is located within the circuit.
```
```suggestion
/// Keeps information about where a bit is located within the circuit.
```

to24toro commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,65 @@
+/// This information includes whether the bit was added by a register,
+/// which register it belongs to and where it is located within it.
+#[derive(Debug, Clone, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct BitInfo {
+    added_by_reg: bool,
+    registers: Vec<BitLocation>,
+}
```
The `BitInfo` struct and its `registers` field may be somewhat ambiguous in conveying their true purpose.It appears that `BitInfo` primarily tracks how bits are associated with registers.

You might consider renaming this `BitInfo` to something more indicative of its focus on bit-register relationships or placements.

However, if you assume how to use it in another way, it is not necessary to rename this field.

raynelfss replied:
You got the intended use correctly. I'm open to renaming this, since it is WiP you can leave any suggestions.

to24toro commented on crates/circuit/src/bit_data.rs:
```diff
@@ -230,3 +235,602 @@ where
 }
+
+#[derive(Clone, Debug)]
+pub struct NewBitData<T: From<BitType>, R: Register + Hash + Eq> {
```
[Q]  Both `BitData<T>` and `NewBitData<T, R>`, which both manage and track bit-related information. However, it’s not entirely clear how they are intended to coexist. Are they meant to be used in parallel for different scenarios(is `BitData<T>` for dag-circuit?), or is there a plan to eventually replace `BitData<T>` with `NewBitData<T, R>`? Any clarification on whether `BitData<T>` is deprecated or if both are needed for distinct reasons would be greatly appreciated.

raynelfss replied:
`NewBitData` here is meant to replace `BitData`, the only reason why I'm keeping the old one around still is because the new replacement methods will need a bigger rewrite of `DAGCircuit`'s methods. Which I plan on doing on a separate PR due to how big it is.

to24toro commented on crates/circuit/src/bit_data.rs:
```diff
@@ -230,3 +235,602 @@ where
+                let res = py_reg.unwrap().bind(py).get_item(bit_info.index())?;
+                self.bits[index_as_usize]
+                    .set(res.into())
+                    .map_err(|_| PyRuntimeError::new_err("Could not set the OnceCell correctly"))?;
```
Instead of simply saying "Could not set the OnceCell correctly", you might include any relevant internal state or the specific condition that led to the failure. This added detail will greatly aid in diagnosing issues.

raynelfss replied:
I agree with this.

to24toro reviewed (commented):
Thank you for sharing this pull request. I understand that this is currently marked as a WIP and a draft, so I was initially hesitant to leave comments. However, I do have a few questions and suggestions that I hope will be helpful as the work progresses.

raynelfss committed:
Fix: Resolve strange behavior of `BitLocations`.
- Fix `BitLocation` updates when a bit or a register is added to use data directly from `CircuitData`.
- Fix incorrect register re-assignment in `QuantumCircuit._from_circuit_data`.
- Re-organize `BitData::py_get_bit_location` to behave more closely to python.

raynelfss committed:
Update bit.rs

Co-authored-by: Kento Ueda <38037695+to24toro@users.noreply.github.com>

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit_u32_sequel

raynelfss committed:
Fix: Add registers correctly in `circuit_to_instruction`.
- Add additional checks to `find_bit()`, to make sure bit exists in the `CircuitData` instance.

raynelfss committed:
Fix: Remaining tests (-1)
- Separate python `add_register` from native method due to conflicting implementation.
- Fix register setter in `NewBitData` to discard old `BitLocations`.
- Update `BitLocations` when setting a new register in `QuantumCircuit`.
- Modify test to include registers when comparing to an equivalency from the `EquivalenceLibrary`.

raynelfss committed:
Fix: incorrect circuit initalization on the rust-side.
- Add one test case to test initialization.

raynelfss committed:
Fix: Regenerate cache if any rust bits are missing.
- Modify the getter methods for the cached lists to correct the cache if it becomes necessary. This to avoid the uninitialization of bits when rust bits are added after requesting access from Python. These methods are now fallible, so a PyResult will always be returned.
- Adapt `CircuitData` and `DAGCircuit` to these changes.

raynelfss committed:
Lint: Remove unused `Qubit` import

raynelfss committed:
Docs: Add header strings in `bit.rs` and `register.rs`

raynelfss committed:
Merge branch 'main' into bit_u32_sequel

raynelfss committed:
Merge branch 'main' into bit_u32_sequel

raynelfss committed:
Fix: Remove changes to the `Interner`
- Remove stale methods `contains` and `get_interned` from `Interner` in favor of the changes introduced by #13752.

raynelfss committed:
Tests: Add rust-side python testing for rust-native circuit creation
- Temporarily fix incorrect null assignment of `Bit` object to indices. A different solution can be achieved if more time is spent.
- Incorrect python circuit creation in Python due to misusage of the bit's `BitInfo` property inside of `BitData`.
- Fix incorrect register and bit creation from `CircuitData::new()`
- Add provisional `qubits_mut()` and `clbits_mut()` methods while we address the issue with `indices`.

raynelfss committed:
Add; Default to rust bits in rust circuit-generating functions.
- Fix error where rust bits wouldn't be initialized in the cache.
- Provisionally fix error in which rust bits wouldn't have valid mapping to python bits. Bits need to be generated before being able to map them to a rust native index. The current solution implements said changes during the `find_bit` function, in which there is an extra validation step. However, this is certain to consume more resources. The ideal solution would be to generate said mapping when obtaining the Python bit with something similar to a `RefCell`. However, `RefCells` are not thread safe and could cause several mutable references to coexist.

raynelfss committed:
Fix: Use `RwLock` for `Bit` indices mappings.
- Re-purpose `NewBitData.indices` to be modifiable even when if a mutable references is not granted by using `RWLock`. This is done to stay consistent with the behavior of `OnceLock` which allows us to initialize bits upon request. We need to make sure to map a bit once it has been initialized. Otherwise, the circuit will have to regenerate this mapping multiple times during runtime.
- Optimize pre-caching of Python bits and registers by using the cache size. Bits and registers are guaranteed to be initialized in ther order they were added/last accessed, so in the case a cache is initialized prematurely, add the missing bits or registers to the back of the cache while following the same index order.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit_u32_sequel

raynelfss committed:
Fix: Modify `QuantumCircuit` copy to perform a smaller amount of copies of the inner `CircuitData`.
- Fix docstrings in `CircuitData` and `BitData`.

raynelfss committed:
Test: Fix circuit_construction test.
- Remove unnecessary mutable reference in `find_bit()`.

raynelfss committed:
Fix: Add more tests
- Register constructors will now accept `Cow<'_, [T]>` to prevent ownership issues.
- `CircuitData::{qregs, cregs}` will now return a slice of registers, instead of an iterator.
- Add duplicate tests for when registers are added to any circuit from rust.
- Add two more rust native tests for circuit construction with registers present.
- Add the ability of turning a float into a `Param` by calling `Into::into`.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit_u32_sequel

raynelfss committed:
Fix: Implement `CircuitData::copy_empty_like()`.

raynelfss committed:
Tests: Add more rust native circuit construction tests.
- Fix incorrect check during `BitData::add_register`.

raynelfss committed:
Docs: Add release note
- Format tests.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit_u32_sequel

1ucian0 commented:
closing in favor of #13860
